File: sim/preassemble.py
#!/usr/bin/env python
# coding: utf-8

# 各命令の行番号と命令番号の対応をとるとともに、ラベルを定数に置換する
# (行番号, 命令番号) というpairを作れば、ラベルは行番号に置き換えは簡単だから処理できそう
# 最初にmappings, labelsをすべて集めて、次にラベルを置換する。2回入力アセンブリを読み込む

def gather(f):
    mappings = []  # list of (行番号, 命令番号)
    labels = []    # list of (ラベル, 行番号)
    l = 0  # line number
    i = 0  # inst number
    for line in f:
        l += 1
        if line.strip()[0] == '#':
            continue
        elif line.endswith(":\n"):
            labels.append((line[:-2], l))
        else:
            mappings.append((l, i))
            i += 1
    return mappings, labels

def label_to_nlin(labels, label):
    for (l, line) in labels:
        if label in l:
            return line
    return -1

def nlin_to_ninst(mappings, i):
    for (nlin, inst) in mappings:
        if i == nlin:
            return inst
    return -1

def label_to_ninst(labels, mappings, label):
    return nlin_to_ninst(mappings, 1 + label_to_nlin(labels, label))

def subst_labels(mappings, labels, f):
    # returns the substituted list of instructions
    ret = []
    l = 0
    for line in f:
        l += 1
        if line.strip()[-1:] == ':':
            continue
        #print(line)
        if line.strip().startswith('jal '):
            n = label_to_ninst(labels, mappings, line.strip()[4:])
            line = 'jal ' + str(n) + '\n'
        elif line.strip().startswith('bc1'): # bc1t/bc1f
            v = line.strip().split(' ')
            n = label_to_ninst(labels, mappings, v[-1]) - nlin_to_ninst(mappings, l) - 1
            line = v[0] + ' ' + str(n) + '\n'
        elif line.strip().startswith('j '):
            n = label_to_ninst(labels, mappings, line.strip()[2:])
            line = 'j ' + str(n) + '\n'
        elif line.strip().startswith('bne ') or line.strip().startswith('beq '):
            v = line.strip().split(' ')
            n = label_to_ninst(labels, mappings, v[-1]) - nlin_to_ninst(mappings, l) - 1
            line = v[0] + ' ' + v[1] + ' ' + v[2] + ' ' + str(n) + '\n'
        elif line.strip()[0] == '#':
            line = line.strip() + '\n'
        elif line.strip().startswith('mv'):
            v = line.strip().split(' ')
            line = 'or ' + v[1] + ' r0 ' + v[2] + '\n'
        elif line.strip().startswith('subi'):
            v = line.strip().split(' ')
            line = 'addi ' + v[1] + ' ' + v[2] + ' -' + v[3] + '\n'
        ret.append(line)
    return ret

def write_labels(path, labels, mappings):
    with open(path, "w") as f:
        buf = []
        for label,_ in labels:
            buf.append(label + ' ' + str(label_to_ninst(labels, mappings, label)) + '\n')
        f.writelines(buf)
    return buf

def write_ninsts(path, mappings):
    with open(path, "w") as f:
        buf = [str(ninst) + ' ' +  str(nlin) + '\n' for (nlin,ninst) in mappings]
        f.writelines(buf)

def halo16(lines, mappings, labels):
    ret = []
    for line in lines:
        while True:
            if 'ha16' not in line and 'lo16' not in line:
                break
            s = line.find('ha16(')
            if s != -1:
                e = line.find(')', s + len('ha16('))
                l = line[s + len('ha16('):e]
                ninst = label_to_ninst(labels, mappings, l)
                num = ninst & 0xf0
                line = line[:s] + str(num) + line[e+1:]
            s = line.find('lo16(')
            if s != -1:
                e = line.find(')', s + len('lo16('))
                l = line[s + len('lo16('):e]
                ninst = label_to_ninst(labels, mappings, l)
                num = ninst & 0x0f
                line = line[:s] + str(num) + line[e+1:]
        ret.append(line)
    return ret

path = 'foo.s'
out = 'piyo.s'

import sys

if __name__ == '__main__':
    #print(len(sys.argv))
    """
    if len(sys.argv) == 3:
        path = sys.argv[1]
        out = sys.argv[2]
    """
    labels = []
    mappings = []
    with open('label.txt') as f:
        for line in f:
            tmp = line[:-1].split(" ") # excluding '\n'
            labels += [(tmp[0], int(tmp[1]))]
    with open('inst.txt') as f:
        for line in f:
            tmp = line[:-1].split(" ") # excluding '\n'
            mappings += [(int(tmp[0]), int(tmp[1]))]
    #with open(path) as f:
    #    mappings, labels = gather(f)
    print(mappings)
    print(labels)
    with open(path) as f:
        ret = subst_labels(mappings, labels, f)
        ret = halo16(ret, mappings, labels)
        with open(out, "w") as writer:
            writer.writelines(ret)
    #write_labels('label.txt', labels, mappings)
    #write_ninsts('inst.txt', mappings)

File: sim/preassemble.cc
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <iostream>
#include <string>
#include <sstream>
#include <fstream>

std::vector<std::string> split(std::string s, std::string delimiter, bool shrink=true);

std::unordered_map<int,int> line_inst_list;
std::unordered_map<int,int> inst_line_list;
std::unordered_map<std::string,int> label_inst_list;

/** line_inst_list, label_inst_list をセット */
void gather(std::string path)
{
  std::ifstream ifs(path);
  if (ifs.fail()) {std::cerr << "File " << path << " cannot be opened. Abort\n"; exit(1);}

  std::string s;
  std::vector<std::string> v;

  int l = 0;  // line number
  int i = 0;  // inst number

  while (1) {
    l++;
    if (!std::getline(ifs, s)) break;
    std::string ss = split(s, "#", false)[0]; // 行の途中のコメントは削除
    v = split(ss, " ");
    if (v.empty()) continue;
    //if (!v[1].compare("")) {std::cout << "Empty (" << l << "): " << s << "\n"; continue;}
    s = v[0];
    v = split(s + "hoge\n", ":", false);
    if (v.size() > 1) label_inst_list.emplace(std::make_pair(v[0], i));
    else {
      line_inst_list.emplace(std::make_pair(l, i));
      inst_line_list.emplace(std::make_pair(i, l));
      i++;
    }
  }
  ifs.close();
}

void subst_labels(std::string infile, std::string outfile)
{
  std::ifstream ifs(infile, std::ios::in);
  if (ifs.fail()) {std::cerr << "File " << infile << " cannot be opened. Abort\n"; exit(1);}
  std::ofstream ofs(outfile, std::ios::out | std::ios::trunc);
  if (ofs.fail()) {std::cerr << "File " << outfile << " cannot be opened. Abort\n"; exit(1);}

  std::string s;
  std::vector<std::string> v;

  int l = 0;  // line number
  while (1) {
    l++;
    if (!std::getline(ifs, s)) break;
    std::string ss = split(s, "#", false)[0]; // 行の途中のコメントは削除
    v = split(ss, " ");
    if (v.empty()) {ofs << s << "\n"; continue;} // comment
    std::vector<std::string> vv = split(s + "foo", ":", false);
    if (vv.size() > 1) {ofs << "###__" << s << "__\n"; continue;} // label declaration
    if (!s.compare("")) {std::cout << "nothing\n"; continue;}

    std::string opcode = v[0];
    std::string str;

    if (!opcode.compare("jal") || !opcode.compare("j")) ofs << opcode << " " << label_inst_list.at(v[1]) << "\t\t\t# " << v[1] << "\n";
    else if (!opcode.compare("bc1f") || !opcode.compare("bc1t")) {
      int n = label_inst_list.at(v[1]) - line_inst_list[l] - 1;
      ofs << opcode << " " << std::to_string(n) << "\t\t\t# " << v[1] << "\n";
    }
    else if (!opcode.compare("bne") || !opcode.compare("beq")) {
      int n = label_inst_list.at(v[3]) - line_inst_list[l] - 1;
      ofs << opcode << " " << v[1] << " " << v[2] << " " << std::to_string(n) << "\t\t\t# " << v[3] << "\n";
    }
    else if (!opcode.compare("mv")) ofs << "or " << v[1] << " r0 " << v[2] <<"\n";
    else if (!opcode.compare("subi")) ofs << "addi " << v[1] << " " << v[2]  << " -" << v[3] <<"\n";
    else if (!opcode.compare("lui")) {
      std::vector<std::string> vv = split(v[2], "ha16(", false);
      if (vv.size() == 1) ofs << s << std::endl;  // line without ha16
      else if (vv.size() == 2) {
        std::string ss = split(vv[1], ")")[0];
        uint32_t n = (uint32_t)label_inst_list.at(ss) & 0xffff0000;
        ofs << opcode << " " << v[1] << " " << std::to_string(n) << "\t\t\t# " << ss << "\n";
      }
      else if (vv.size() > 2) {
        std::cerr << "Too many halo16 in a line(" << l << ". Abort.\n"; exit(1);  // unexpected
      }
    }
    else if (!opcode.compare("ori")) {
      std::vector<std::string> vv = split(v[3], "lo16(", false);
      if (vv.size() == 1) ofs << s << std::endl;  // line without lo16
      else if (vv.size() == 2) {
        std::string ss = split(vv[1], ")")[0];
        uint32_t n = (uint32_t)label_inst_list.at(ss) & 0xffff;
        ofs << opcode << " " << v[1] << " " << v[2] << " " << std::to_string(n) << "\t\t\t# " << ss << "\n";
      }
      else if (vv.size() > 2) {
        std::cerr << "Too many halo16 in a line(" << l << ". Abort.\n"; exit(1);
      }
    }
    else ofs << s << std::endl;
  }
  ifs.close();
  ofs.close();
}

void output_inst(std::unordered_map<int,int> line_inst_list, std::string path = "inst.txt")
{
  std::string s = "";
  std::ofstream ofs(path);
  if (ofs.fail()) {std::cerr << "File " << path << " cannot be opened. Abort\n"; exit(1);}
  for (auto it : line_inst_list) {
    s += std::to_string(it.second);
    s.append(" ");
    s += std::to_string(it.first);
    s.append("\n");
  }
  ofs << s;
}

void output_labels(std::unordered_map<std::string,int> label_inst_list, std::string path = "label.txt")
{
  std::string s = "";
  std::ofstream ofs(path);
  if (ofs.fail()) {std::cerr << "File " << path << " cannot be opened. Abort\n"; exit(1);}
  for (auto it : label_inst_list) {
    s += it.first;
    s.append(" ");
    s += std::to_string(it.second);
    s.append("\n");
  }
  ofs << s;
}

/**--- std::string をdelimiterで分割してstd::vectorで返す ---*/
std::vector<std::string> split(std::string s, std::string delimiter, bool shrink)
{
  std::vector<std::string> v;
  std::string::size_type pos = 0;

  if (!delimiter.compare("")) v.push_back(s);
  else {
    while (1) {
      if (shrink) {
        while (1) {
          if (pos >= s.length()) break;
          if (s.find(delimiter, pos) == pos) pos += delimiter.length();
          else break;
        }
      }
      if (pos >= s.length()) break;
      std::string::size_type tmp = s.find(delimiter, pos);
      if (tmp == std::string::npos) {v.push_back(s.substr(pos, tmp-pos)); break;}
      v.push_back(s.substr(pos, tmp-pos));
      pos = tmp + delimiter.length();
    }
  }
  return v;
}

void test()
{
  std::string s = "    #aaa";
    s = split(s, "#", false)[0]; // 行の途中のコメントは削除
    std::vector<std::string> v = split(s, " ");
    if (v.empty()) {std::cout << "empty\n";}
    std::vector<std::string> vv = split(s + "foo", ":", false);
    if (vv.size() > 1) {std::cout << "vv " << vv.size() << std::endl;}
    if (!s.compare("")) std::cout << "\"\"" << std::endl;
  for (auto it : v) {
    std::cout << it << std::endl;
  }
  //std::cout << s << std::endl;
}

int main(int argc, char **argv)
{
  std::string in = "foo.s";
  std::string out = "piyo.s";

  if (argc >= 2) in = argv[1];
  if (argc == 3) out = argv[2];
  if (argc > 3) {std::cerr << "USAGE: " << argv[0] << "{{in_file}} {{preassembled_file}}\n"; exit(1);}

  std::cout << in << " ==> " << out << std::endl;
  gather(in);
  output_inst(line_inst_list);
  output_labels(label_inst_list);
  subst_labels(in, out);
  //test();

  return 0;
}
